Replace SQLAlchemy Migrate with Alembic

jdavcs opened the pull request:
**NOTE:** This branch has been superseded by #13513. However, the description of the migration system provided with this PR is accurate and can be used as implementation reference for #13513.

---

Ref #10369.
Supersedes #12869.

This is a clean branch with a cleaned-up commit history; some commits are large, but all changes are logically grouped together. (the commit history in #12869 shows most of the previous iterations)

## Migrations system overview

Galaxy's data model is split into the galaxy model and the install model. These two models may be persisted in one combined database or two separate databases. To accommodate this setup, as well as a seamless transition from a combined database to separate databases and vice-versa, the proposed new Alembic-based migrations system is built on top of branches.

A branch, in this context, is a versioning lineage (or migration stream) that starts at a common root (base revision) and represents some part of the data model. These branches are permanent, are identified by labels, and may share the same Alembic version table (if they share the same database; otherwise, each database has its own version table). 

We use 2 branches: `gxy` (galaxy) and `tsi` (tool shed install). These IDs were selected for their brevity and visual consistency: they are referenced throughout the system as version directory names and branch labels (by Alembic), and model identifiers. In addition, the term "install" in "install model" may be ambiguous in some contexts (it's easy to mistake it for a verb, etc.).

Each branch has its own version history, represented by revision scripts located in the branch version directory (e.g. revisions for the galaxy model are located in the `gxy` version directory, whereas revisions for the install model are located in the `tsi` version directory). The version directories contain labeled base revisions, which were created with the corresponding branch labels. When we create a new revision, we only need to indicate the branch label: the system will know where to place the new revision file.

Alembic requires its own configuration file (`alembic.ini`); we use it for required settings only: the location of the alembic script and the version directories, and the default Alembic logging configuration. All other configuration information, including how the two models are persisted, is programmatically set at run time based on Galaxy's own configuration, which remains the one source of truth. There are no changes to how we indicate separate vs. combined database storage: if the `install_database_connection` option is set to a value other than the value of `database_connection`, two databases will be used instead of a combined database. If the setting is changed for an existing database, our system will know how to handle this (see tests for such scenarios).

See relevant Alembic documentation: 
https://alembic.sqlalchemy.org/en/latest/branches.html#working-with-branches
and especially here:
https://alembic.sqlalchemy.org/en/latest/branches.html#working-with-explicit-branches
https://alembic.sqlalchemy.org/en/latest/branches.html#working-with-branch-labels
https://alembic.sqlalchemy.org/en/latest/branches.html#working-with-multiple-bases

## Directory structure

```
lib/galaxy/model/migrations/
├── alembic
│   ├── versions_gxy
│   │   └── e7b6dcb09efd_create_gxy_branch.py
│   ├── versions_tsi
│   │   └── d4a650f47a3c_create_tsi_branch.py
│   ├── env.py
│   └── script.py.mako
├── README.md
├── __init__.py
├── alembic.ini
└── scripts.py
```
`versions_*`: version directories, containing revision scripts; filenames include a GUID-based revision identifier, assigned by Alembic. Ordering is set via directives inside the revision scripts (`down_revision`)

`env.py`: invoked by Alembic whenever a command is executed. Its primary purpose is to determine the target database (based on the indicated branch) and configure the migration environment.

`script.py.mako`: the template file used to generate revisions.

`README.md`: basic user documentation on how to use the new migrations system.

`__init__.py`: system core (see _Main classes and entry points_).

`alembic.ini`: Alembic configuration file

`scripts.py`: helpers used for invoking Alembic via script (`create_db/migrate_db`)

#### Main classes and entry points

* `verify_databases`, `verify_databases_via_script`: entry points

* `AlembicManager`: handles Alembic interactions with one database; serves as a facade to a subset of Alembic commands.

* `DatabaseStateCache`: holds a snapshot of database state. This is an abstraction layer that simplifies the code and  reduces (1 vs. 4) the number of calls to the database.

* `DatabaseStateVerifier`: executes the core migration algorithm.

## Migration algorithm

When invoked via `create_db.sh` script or as part of Galaxy's startup process (in `config/__init__.py`), the system follows the same algorithm. (Unlike `create_db.sh`, the `migrate_db.sh` script is a thin wrapper around Alembic, so it does not verify the database state; it determines the target branch, retrieves relevant configuration values and invokes Alembic.)

We start at one of the two entry points; then, subsequently, create and run a `DatabaseStateVerifier` for the `gxy` model, then for the `tsi` model. The system checks if the database exists and is non-empty, and then determines whether it is up-to-date, and if not, whether it can be upgraded automatically. The algorithm tries to minimize database calls by using a cache (`DatabaseStateCache`) and by taking shortcuts when possible (e.g. `is_new_database` will be set if while processing the first model a new or an empty database was initialized).

The algorithm makes "an educated guess" only in one case. If within an existing, nonempty database that is combined, the `gxy` model is up-to-date, but the `tsi` model is not, we cannot determine the state of the `tsi` model tables with absolute certainty. Such an edge might happen for several reasons (see inline comments in [`TestPartiallyUpgradedCombinedDatabase`](https://github.com/galaxyproject/galaxy/blob/8e11b98edfcce908a016059ba919c9696d92d81b/test/unit/data/model/migrations/test_migrations.py#L607) for a detailed description). We cannot determine the state of this `tsi` model because we do not store the SQLAlchemy Migrate version of the install model in a combined database. So, the reasoning goes like this: if there are 
no `tsi` tables in the database, we simply treat it as a new install; but if there is at least one table, we assume it is the same version as `gxy` (if `gxy` was manually upgraded, we assume the same for `tsi`). I think, there's nothing else we can reasonably do, other than fail unconditionally, which seems unreasonable in this case.

The diagram below describes the algorithm.  

(**UPDATED on 8/21/22**: same logic; better image from GCC 2022 talk)
![image](https://user-images.githubusercontent.com/6132258/185800540-37bb9be2-9d11-4194-be1a-69c3a963741e.png)

(**UPDATED ON 2/18/22**; see changes to the paths "has alembic table? > no" and "is sqlalchemy migrate version latest? > yes")
![20220218_195436](https://user-images.githubusercontent.com/6132258/154779347-aad068be-67c2-4314-95c6-abee62a46d8e.jpg)

Previous version:

![alembic_board](https://user-images.githubusercontent.com/6132258/148877964-c9e0a2a0-55fb-423f-840c-f564dc934fb8.jpg)

## Misc. implementation notes:
* Old migration scripts have been discarded: we either start from an empty database, or from a database that has been upgraded to the last SQLAlchemy Migration (version 180), at which point an upgrade can be ran automatically. 

* We don't automatically upgrade the database; we do so only if the `database_auto_migrate` option has been set`. Otherwise, the systems raises an error with an informative message.

* SQLAlchemy Migrate has not been deleted from Galaxy's Python dependencies list because it is still used for Tool Shed. However, SQLAlchemy Migrate is not compatible with SQLAlchemy 2.0, so moving Tool Shed to Alembic as well seems a reasonable step, given Galaxy's upcoming migration to SQLAlchemy 2.0. Meanwhile, as a temporary fix, I have created TS-specific scripts to create and migrate the Tool Shed databse using SQLAlchemy Migrate.

* The `GALAXY_ALLOW_FUTURE_DATABASE` setting has been removed. Revision ordering is no longer determined by a sequence of ascending integers, so determining whether a given revision head stored in the database is a "future" version is no longer possible (if it's more than one version removed from the head in the code). Since this was used for development instances only, an alternative solution is to simply set `check_migrate_databases=False`. Another alternative is to keep the new revision scripts in the versions directories, so that the system can identify the head stored in the database as a valid revision. Although, I think, if one is knowingly running a database whose version is out of sync with the code base, verifying the database state programmatically makes little sense, so just unset `check_migrate_databases`.

## Discarded approaches

I considered multiple alternatives, primarily these:
- Use 2 alembic directories (2 sets of everything)
- Use 2 migration environments (galaxy and install), use `--name` argument to distiguish between the two
(ref: https://alembic.sqlalchemy.org/en/latest/cookbook.html#run-multiple-alembic-environments-from-one-ini-file)
- Variations of the above

Reasons for discarding (apply to some or all of the discarded options):
- Applicable to completely independent lineages: not the case with galaxy/install
- Impossible to set dependencies between branches: we might need that
- Lots of duplicate code: we only need one `env.py` and one `mako` template (with 2 alembic directories, we get duplicate `alembic.ini` as well)
- Can't run without the `--name` argument, so for a combined database would need a 3 values: galaxy, install, both - which doesn't feel right for multiple reasons
- Main reason: not suitable for our core use case: although we have 2 distinct models, they may be mapped to one ***OR*** two databases, and **we must accommodate both**. Furthermore, setups may change: 2 >> 1 and 1 >> 2. I ran into multiple issues with this point, and was unable to find a reasonably clean implementation. The main issue was targeting commands at one or the other or both with different database setups.

#### Discarded migration algorithm
In the current version, the gxy and tsi models are verified in sequence. In a previous version, they were verified at the same time. That saved a few trips to the database, but the logic was much less straightforward, the design was awkward, and adding/removing additional models/databases would have required rewriting the algorithm. It wasn't a good design and the minimal optimization gains, in my opinion, did not justify the lack of clarity and flexibility.
https://github.com/ic4f/galaxy/blob/7a34c3388feaef05b2d0de73844ddd855743cfcc/lib/galaxy/model/migrations/__init__.py

## Usage

See [README.md](https://github.com/galaxyproject/galaxy/blob/78d7786fa74c51b320d522f1a124a2331f08bc0e/lib/galaxy/model/migrations/README.md)

## Testing

1. Unit tests + infrastructure located in `test/data/model/migrations`. 
2. An additional github workflow has been added that runs all the migrations tests on postgresql
9c96664a70910f4009b10d80fc249d4e386fb1cb

Following is a description of the main parts of the testing infrastructure.

### `conftest.py`

Contains the following:
1. Composable "atomic" MetaData objects (implemented as fixture factories), containing one table object.
Used to compose representations of database state.  Example:

```python
@pytest.fixture
def gxy_table1():
    def make_table(metadata):
        return sa.Table('gxy_table1', metadata, sa.Column('id', sa.Integer, primary_key=True))
    return make_table
```

2. MetaData objects containing one or more tables and representing database state. Example:

```python
@pytest.fixture
def metadata_state3_combined(
    gxy_table1, gxy_table2, tsi_table1, tsi_table2, sqlalchemymigrate_table
):
    metadata = sa.MetaData()
    gxy_table1(metadata)
    gxy_table2(metadata)
    tsi_table1(metadata)
    tsi_table2(metadata)
    sqlalchemymigrate_table(metadata)
    return metadata
```
The object above is composed of 5 tables defined in fixture factories.

These MetaData objects are used to load a database with a given state. For example, the following test function would load this metadata into the provided database:

```python
def test_state3_combined(engine, metadata_state3_combined):
    with engine.connect() as conn:
        metadata_state3.create_all(bind=conn)
```
The database states are generated in fixtures (see bottom of `test_migrations.py` + inline comments for more details).
Each state has 3 versions, distinguished by suffix:
- `_gxy` (galaxy database that holds gxy* and migration version tables)
- `_tsi` (tool_shed_install database that holds tsi* and migration version tables)
- `_combined` (combined database that holds gxy*, tsi*, and migration version tables)

The following states are represented:

- State 1: Non-empty database, no version table.
          (Most ancient state)
- State 2: SQLAlchemy Migrate version table added.
         (Oldest state versioned with SQLAlchemy Migrate)
- State 3: New table added.
         (Last (most recent) state versioned with SQLAlchemy Migrate)
- State 4: Alembic version table added.
         (Oldest state versioned with Alembic)
- State 5: SQLAlchemy Migrate version table removed.
         (Oldest state versioned with Alembic that does not include the SQLAchemy Migrate version table)
- State 6: New table added.
         (Most recent state versioned with Alembic. This is the current state)
- Additional edge case: gxy at state3, tsi has no SQLAlchemy Migrate table.

(State 0 assumes an empty database, so it needs no state fixtures.)

### `common.py`

Contains helpers (context managers and fixtures) for generating database connection URLs, creating and dropping databases, and creating self-disposing engines. Handles both sqlite and postgresql databases. (mysql not supported yet)

### `versions/`

Contains revision scripts used in some of the tests.

### `test_migrations.py`

Contains:
- tests for AlembicManager
- tests for DatabaseStateCache
- tests for DatabaseVerifier (which handles the migration process) in various contexts (different database state and databse setup: see inline comments for details)
- database fixtures loaded with metadata from `common.py` 
- tests for these database fixtures (verifying that they have the expected state)
- various helper fixtures and functions + tests for helper functions

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:

Following is a testing plan I wrote for myself and followed to manually test the system. It doesn't have to be repeated in its entirety.

To create a revision:
1) `./migrate_db.sh revision --head=gxy@head -m "revision message..."` (use "gxy" or "tsi")
2) fill-in upgrade/downgrade empty functions in the generated revision script
(for examples, see `test/data/model/migrations/versions/db1`)

For upgrades/downgrades: also try: append `--sql` to do a dry run

To upgrade to head:
1) `./migrate_db.sh upgrade gxy@head`
2) ***verify changes in database***

To upgrade/downgrade to a specific revision:
1) `./migrate_db.sh upgrade rev#`  (for rev# use several starting chars from the revision id)
2) ***verify changes in database***

**1. Initial state: no database, separate database setup**
Prep: delete old databases, delete old revisions, edit galaxy.yml

1.1. gxy, sqlite database
- gxy-rev1: create table1
- `upgrade` gxy@head
- gxy-rev2: create table2
- gxy-rev3: create table3
- `upgrade` to gxy-rev2 (use revision identifier)
- `upgrade` to gxy-rev3 (use head)
- `downgrade `to gxy-rev1 (use revision identifier)
- `downgrade` to gxy-base (use base)
- `current`: should see nothing
- `upgrade`: use alembic relative upgrade: gxy@+2
- `downgrade`: use alembic relative downgrade: gxy@-2

1.2. tsi, sqlite database
- tsi-rev1: create table1
- `upgrade heads`
- tsi-rev2: create table2
- tsi-rev3: create table3
- `upgrade` to tsi-rev2 (use +1)
- `upgrade` to tsi-rev3 (use head)
- `downgrade` tsi@-1
- `downgrade` to tsi@base

1.3. repeat steps 1.1, 1.2 with postgresql

**2. Initial state: no database, combined databases setup**
Prep: delete old databases, delete old revisions, edit galaxy.yml

2.1. gxy, sqlite
- repeat steps from 1.1

2.2. tsi, sqlite
- repeat steps from 2.1

2.3. repeat steps 2.1, 2.2 with postgresql

**3. Initial state: old database (create state in another git branch), separate databases setup**

3.1. sqlite, run Galaxy:
- no SQLAlchemy Migrate table in gxy: verify fail
- old SQLAlchemy Migrate table in gxy: verify fail
- new SQLAlchemy Migrate table in gxy, no automigrate: verify fail
- new SQLAlchemy Migrate table in both, automigrate: verify upgrade

3.2. postgres, run Galaxy:
- no SQLAlchemy Migrate table in gxy: verify fail
- old SQLAlchemy Migrate table in gxy: verify fail
- new SQLAlchemy Migrate table in gxy, no automigrate: verify fail
- new SQLAlchemy Migrate table in both, automigrate: verify upgrade

**4. Initial state: old database (create state in another git branch), separate databases setup, sqlite, gxy is with alembic and current**
Run Galaxy:
- no SQLAlchemy Migrate table in tsi: verify fail
- old SQLAlchemy Migrate table in tsi verify fail
- new SQLAlchemy Migrate table in tsi, no automigrate: verify fail

**5. Initial state: old database (create state in another git branch), combined database, sqlite**
Run Galaxy:
- no SQLAlchemy Migrate table: verify fail
- old SQLAlchemy Migrate table: verify fail
- new SQLAlchemy Migrate table, no automigrate: verify fail
- new SQLAlchemy Migrate table, automigrate: verify upgrade

**6. Initial state: old database (create state in another git branch), combined database, postgres**
Run Galaxy:
- no SQLAlchemy Migrate table: verify fail
- old SQLAlchemy Migrate table: verify fail
- new SQLAlchemy Migrate table, no automigrate: verify fail
- new SQLAlchemy Migrate table, automigrate: verify upgrade

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

jdavcs committed:
Add alembic dir to ignore list in run_tests

Otherwise alembic/env.py gets executed out of context at test collection
stage (via tox -e unit >> run_tests.sh), which raises an AttributeError
(alembic.context is not set in this case)

jdavcs committed:
Add alembic dir to ignore list in package tests

Otherwise alembic/env.py gets executed out of context at test collection
stage (via tox -e unit >> run_tests.sh), which raises an AttributeError
(alembic.context is not set in this case)

jdavcs committed:
Move engine_options default arg initialization to engine_factory

To eliminate the repetitive `engine_options = engine_options or {}` in
multiple places.

jdavcs committed:
Remove migrations test from first starup workflow

This tests sqlalchemy migrate which is being replaced.

jdavcs committed:
Add workflow for migration tests

jdavcs committed:
Run tests under Python 3.7

jdavcs committed:
Add alembic to dependencies

jdavcs committed:
Add testing infrastructure

jdavcs committed:
Add alembic infrastructure

jdavcs committed:
Move triggers out of migrate directory

(does not include an import in mapping.py: to be added in a separate
commit, grouped with other edits in mapping.py)

jdavcs committed:
Integrate migrations design into config and model code

jdavcs committed:
Handle error when revision not found; add test case

jdavcs committed:
Avoid extra trip to db via lazy load of alembic version heads

jdavcs committed:
Use correct db options for both engines; resolve todos

jdavcs committed:
Fix mypy error

jdavcs committed:
Add informative log/error messages; add tests for new methods

jdavcs committed:
Drop config.database_create_tables (not used)

jdavcs committed:
Drop app.new_installation (not used)

jdavcs committed:
Don't use triple-quoted strings as comments in tests

Ref #13007

jdavcs committed:
Setup infrastructure for running db scripts

jdavcs committed:
Add create/migrate scripts for tool shed

jdavcs committed:
Add create/migrate scripts for Alembic dbs

Also add exec permissions to create/migrate python scripts.
(not sure this is required, adding for consistency with previous
version)

+ A few fixes for TS scripts

jdavcs committed:
Fix integration test (test only relevant for TS)

jdavcs committed:
Fix mypy error

jdavcs committed:
When overriding, override.

Well, that was dumb..

jdavcs committed:
Update data package (requirements, packages, includes)

Note: manage_db entry point needs to be edited in model.orm.scripts.

jdavcs committed:
Move helper function out of migrate into triggers

jdavcs committed:
Remove galaxy.model.migrate files

jdavcs committed:
Remove galaxy.model.tool_shed_install.migrate files

jdavcs committed:
Move is-one-db check into model.database_utils

jdavcs committed:
Update data package entry point in orm.scripts.py

jdavcs committed:
Remove outdated parts from orm.scripts.py

jdavcs committed:
Update db index testing: test both models, gxy and tsi

Also, remove unused env var

jdavcs committed:
Add a comment on sqlalchemy-migrate to pyproject.toml

jdavcs committed:
Address existing mypy error (see note)

This is a (possible) bug that has existed on dev. To expose it, simply
change this call to init_models_from_config
(https://github.com/galaxyproject/galaxy/blob/release_21.09/lib/galaxy/config/__init__.py#L1302)
to a call directly to mapping.init(). The result will be 6 mypy errors.

The reason is that the mapping.init() function
(https://github.com/galaxyproject/galaxy/blob/release_21.09/lib/galaxy/model/mapping.py#L79)
specifies its return type as GalaxyModelMapping - which results in
several errors of this type:

lib/galaxy/app.py:178: error: Definition of "model" in base class "MinimalApp" is incompatible with definition in base class "ConfiguresGalaxyMixin"

When we call an intermediary function (that does not have a return type
specified), mypy is happy.

This should be addressed in a separate issue/PR. Removing the return
type is the simplest (temporary) solution.

jdavcs committed:
Dispose of engines when invoked via script

jdavcs committed:
Add type hints to migrations modules

jdavcs committed:
Type-ignore alembic type hint bug

jdavcs committed:
Add type hints to common module in migration tests

jdavcs committed:
Remove circular dependency between migrations and config

jdavcs committed:
Add migrations README

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -0,0 +1,70 @@
+
+1. Backup your database(s).
+
+2. Verify that your database is at the latest SQLAlchemy Migrate version. If you have a combined database, it should be version 180 (check the `migrate_version` table). If you have separate galaxy model and install model databases, your galaxy version should be 180, and your install model version should be 17. 
```
180 was added in dev, so people can't jump to a new release and update. Should we provide the 180 migration out of the box ? I think that's a good test and a jumping off point for devs that need to change/add/remove tables.

jdavcs replied:
Will do.

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -0,0 +1,70 @@
+
+2. Verify that your database is at the latest SQLAlchemy Migrate version. If you have a combined database, it should be version 180 (check the `migrate_version` table). If you have separate galaxy model and install model databases, your galaxy version should be 180, and your install model version should be 17. 
+
+3. If your database is not current, you need to update it first. For that, you'll need to check out a previous version of the code base: use commit 7328ca0da9977bc31cd074464a2e59411a5cdcf8 or later, but prior to [TODO: insert Alembic commit]. Then you can run the old `manage_db.sh` script to bring your database(s) up to date. Once your database has the latest SQLAlchemy Migrate version, you can c ...
```
That's too complicated. People should be able to go to the next release and run `./manage_db.sh upgrade`

jdavcs replied:
Maybe I phrased it in a confusing way. The idea is (as in my comment below), to use 22.01, you need to be current - i.e., you need to be on 21.09. You cannot "jump" over releases. If you are not - just make sure you are current on 21.09, then upgrade. If this is acceptable, I'll rephrase the readme.

jdavcs replied:
Done and done: they can go to the next release and run `./manage_db.sh upgrade` (provided they were up-to-date on the previous release; otherwise they just run the same command on that release). And the steps in README have been updated to be simple and straightforward.

mvdbeek replied:
Excellent! I'll give the test upgrade another go

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -0,0 +1,70 @@
+
+3. If your database is not current, you need to update it first. For that, you'll need to check out a previous version of the code base: use commit 7328ca0da9977bc31cd074464a2e59411a5cdcf8 or later, but prior to [TODO: insert Alembic commit]. Then you can run the old `manage_db.sh` script to bring your database(s) up to date. Once your database has the latest SQLAlchemy Migrate version, you can c ...
+
+4. If you want Alembic to upgrade your database automatically, you can set the [`database_auto_migrate`](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/config_schema.yml#L170) configuration option and simply start Galaxy. Otherwise, use the `migrate_db.sh` script.
```
Does that work if you're not on the latest migration ?

jdavcs replied:
It won't work if you are not on the latest SQLAlchemy Migrate migration. For that to work, the system would have to first use SQLAlchemy Migrate to upgrate to *that* latest version, and after that pass the control onto Alembic. However, SQlAlchemy Migrate won't run under SA 2.0 (it would break once we enable the future flags in our SA 1.4 >> 2.0 migration process). 
If you look at the whiteboard diagram above, the check for `database_auto_migrate` happens only in one place (bottom right corner) - and at that step, the database is either under alembic version control (in some form - i.e., it varies), or it has the latest SQLAlchemy Migrate version. So we ask to *first* ensure the database is current for the previous release, which, I hope, is a reasonable prerequisite.

mvdbeek replied:
Hmm .... so we know there's a lot of "dependency-rot" happening over the years, and this does mean that ideally 21.09 remains installable for ~ 3-5 years ... which given past performance seems unlikely. Maybe we should create a docker image that can upgrade Galaxy to these points, for someone needing to make the jump in the future.

jdavcs replied:
I'm probably not picturing this correctly.. If we have an instance < 22.01, that instance still runs on the old sqlalchemy migrate code, and sqlalchemy 1.4. So the upgrade path would be the same:
1. `./manage_db.sh upgrade --version=179` (migration handled by sqlalchemy migrate; code base is on SA 1.4)
2. check out new release
3. `./migrate_db.sh upgrade heads` (migration handled by alembic; code base is on SA 1.4 or 2.0)

So, if someone needs to jump from an older (previously supported)  version to 22.01 or beyond, I don't see a problem: they won't be in a situation when the db is under sqlalchemy migrate while their code base is using sqlalchemy 2.0. Or am I missing a use case?

mvdbeek replied:
Some instances get an update every 2 years (or even less often). So 21.09 (the last one that can migrate to 179) needs to remain installable. That could be a challenge. IOW try installing 17.09, I believe you'll have a hard time doing that.

jdavcs replied:
***Very good to know!*** I'm trying this now.

jdavcs replied:
17.09 is a problem for unrelated reasons (some psycopg2-related issue with postgres, and sqlite has [its own issues](https://github.com/galaxyproject/galaxy/issues/7159#issuecomment-450653080)). 
But the next one, 18.01 works (almost) fine. Here's what I did: 
- verified it runs, upgraded codebase to 21.09, recreated the venv with python3, then `manage_db.sh upgrade` - which gives us a healthy 21.09. 
- Then checkout next release (current alembic branch), again `manage_db.sh upgrade` - and that leaves the install model not upgraded. Of course if you run `manage_db.sh upgrade install` it takes care of it. (and, of course, `run_alembic.sh upgrade heads` upgrades both)

The reason is that "default database" in manage_db.sh in the previous version was rather ambiguous: it meant "whatever is in the galaxy database". So, if we were using separate databases, that would be the galaxy model and the install model; but if `install_database_connection` was set to a different db, that would be the galaxy model only. The new version doesn't make such guesses; instead, it requires an explicit indication what to upgrade: galaxy's head, install's head, or heads - for both.

I'll modify the current logic of `manage_db.sh` to match the logic of the previous version: running `manage_db.sh upgrade` will upgrade both models if it's a combined database, and only the galaxy model if there's a separate install database.

jdavcs replied:
OK, this should be fixed. Works as expected: start with 18.01, then `./manage_db.sh upgrade`, then checkout 21.09 (rebuilding .venv), then `./manage_db.sh upgrade`, then checkout alembic branch (aka 22.01), and again `./manage_db.sh upgrade`. 

Along the way discovered a regression: upgrading from dev raised an error when executing the `vault` table revision script (in dev that table was added in db version 180, but not in 21.09). So that's fixed in 6ffed8f.

And a regression on this regression, if we were to:
step 1: upgrade from dev to 22.01 (vault table not created since it's already there)
step 2: downgrade from 22.01 back to dev (vault table dropped)
step 3: in dev, downgrade to a previous db version > error! (trying to drop nonexistent table)

Because after step 2, we'd end up with the current SQLAlchemy Migrate db version 180, which assumes the existence of the `vault` table (which was dropped in step 2 by Alembic!). I'll fix this on dev in a separate PR.

jdavcs committed:
Add prev. version 180 as new revision.

Reason: 180 was added in dev, so it was not in 21.09.

jdavcs committed:
Preserve manage_db.sh, but do not hide alembic

- Preserves usage of legacy script while not hiding the Alembic script.
    (see manage_db.sh and run_alembic.sh)
- Adds documentation (see top of script files)
- Includes misc. relevant refactorings
- Includes unit tests for testing legacy script handling

jdavcs committed:
Update README: rename script, add legacy script, simplify upgrade steps.

jdavcs committed:
Modify behavior of legacy wrapper, add tests, refactor

If the database is combined, running `./manage_db.sh upgrade` should upgrade
BOTH models: gxy and tsi. However, if tsi uses a separate database, only
the galaxy model should be updated.

Also, see this for context: https://github.com/galaxyproject/galaxy/pull/13108#discussion_r789108576

See inline comments for more details.

The new tests are examples of new behavior.

All refactoring is directly related to the changes in this commit.

jdavcs committed:
Check for table existence for vault (see note)

This was the last revision to the database that was handled with
SQLAlchemy Migrate prior to the move to Alembic (db version 180).
However, it was added to the dev branch after 21.09 was released, but
before 22.01. Therefore, this table should be created when upgrading
from 21.09, but not from dev.

jdavcs committed:
Stub out config call in test_scripts; fixes package test

dannon commented on manage_db.sh:
```diff
@@ -1,15 +1,49 @@
-find lib/galaxy/model/migrate/versions -name '*.pyc' -delete
-python ./scripts/manage_db.py $@
+for i; do :; done
+if [ "$i" == "tool_shed" ]; then
```
```suggestion
if [ "$i" = "tool_shed" ]; then
```

dannon replied:
Is this correct?  My bash-fu is dusty, but I think you need to use a single equals in [ ] for equality.

jdavcs replied:
You are right. My Bash Pocket Reference says that singly `=` should be used with `test` and `[ ]` for compatibility with POSIX and other shells. Thanks!

dannon commented on manage_db.sh:
```diff
@@ -1,15 +1,49 @@
+    python ./scripts/migrate_toolshed_db.py $@ tool_shed
+else
+    find lib/galaxy/model/migrations/alembic -name '*.pyc' -delete
+    python ./scripts/manage_db_adapter.py --alembic-config "$ALEMBIC_CONFIG" $@
```
```suggestion
    python ./scripts/manage_db_adapter.py --alembic-config "$ALEMBIC_CONFIG" "$@"
```

dannon commented on manage_db.sh:
```diff
@@ -1,15 +1,49 @@
-python ./scripts/manage_db.py $@
+for i; do :; done
+if [ "$i" == "tool_shed" ]; then
+    python ./scripts/migrate_toolshed_db.py $@ tool_shed
```
```suggestion
    python ./scripts/migrate_toolshed_db.py "$@" tool_shed
```

jdavcs committed:
Update manage_db.sh ($@ > "$@")

Co-authored-by: Dannon <dannon.baker@gmail.com>

jdavcs committed:
Update manage_db.sh ($@ > "$@")

Co-authored-by: Dannon <dannon.baker@gmail.com>

jdavcs committed:
Update manage_db.sh (use single `=` in `[ ]`)

Co-authored-by: Dannon <dannon.baker@gmail.com>

jdavcs committed:
Fix import order in config

jdavcs committed:
Apply black, isort

jdavcs committed:
Remove old migrate files (fix rebase)

jdavcs committed:
Change latest db version to 180

jdavcs committed:
Adapt manage_db.py for usegalaxy playbook; add tests

jdavcs committed:
Fix some of the package tests

jdavcs committed:
Fix path for package tests

jdavcs committed:
Print result of script execution

jdavcs committed:
One engine for 2 dbs logically equiv to 2 refs to same engine

jdavcs committed:
Change handling of unrecognized revisions; improve log messages

jdavcs committed:
Remove vault table revision (part of 22.01)
